eqcorrscan/tests/correlate_test.py

```diff
@@ -2,112 +2,395 @@
+    except AttributeError:  # test does not belong to a class
+        return
+    else:
+        dash = '-' * 70
+        print('\nstarting tests contained in class %s\n%s' % (cls_name, dash))
+
+
+# array fixtures
+
+starting_index = 500
+
+
+@pytest.fixture(scope='module')
+def array_template():
+    """
+    return a set of templates, generated with randomn, for correlation tests.
+    """
+    return random.randn(200, 200)
+
+
+@pytest.fixture(scope='module')
+def array_stream(array_template):
+    """
+    Return a stream genearted with randomn for testing normxcorr functions
     """
-    Run core tests
+    stream = random.randn(10000) * 5
+    # insert a template into the stream so cc == 1 at some place
+    ar = array_template[0]
+    stream[starting_index: starting_index + len(ar)] = ar
+    return stream
+
+
+@pytest.fixture(scope='module')
+def pads(array_template):
     """
-    unittest.main()
+    return an array of zeros for padding, matching templates len.
+    """
+    return np.zeros(array_template.shape[0], dtype=int)
+
+
+@pytest.fixture(scope='module')
+def array_ccs(array_template, array_stream, pads):
+    """  use each function stored in the normcorr cache to correlate the
+     templates and arrays, return a dict with keys as func names and values
+     as the cc calculated by said function"""
+    out = {}
+
+    for name in list(corr.XCOR_FUNCS.keys()):
+        func = corr.get_array_xcorr(name)
+        cc, _ = time_func(func, name, array_template, array_stream, pads)
+        out[name] = cc
+    return out
+
+
+# stream fixtures
+
+
+@pytest.fixture(scope='module')
+def multichannel_templates():
+    """ create multichannel templates """
+    return generate_multichannel_templates()
+
+
+@pytest.fixture(scope='module')
+def multichannel_stream():
+    """ create a multichannel stream for tests """
+    return generate_multichannel_stream()
+
+
+# a dict of all registered stream functions (this is a bit long)
+stream_funcs = {fname + '_' + mname: corr.get_stream_xcorr(fname, mname)
+                for fname in corr.XCOR_FUNCS.keys()
+                for mname in corr.XCORR_STREAM_METHODS
+                if fname != 'default'}
+
+
+@pytest.fixture(scope='module')
+def stream_cc_output_dict(multichannel_templates, multichannel_stream):
+    """ return a dict of outputs from all stream_xcorr functions """
+    # corr._get_array_dicts(multichannel_templates, multichannel_stream)
+    out = {}
+    for name, func in stream_funcs.items():
+        cc_out = time_func(func, name, multichannel_templates,
+                           multichannel_stream)
+        out[name] = cc_out
+    return out
+
+
+@pytest.fixture(scope='module')
+def stream_cc_dict(stream_cc_output_dict):
+    """ return just the cc arrays fro mthe stream_cc functions """
+    return {name: result[0] for name, result in stream_cc_output_dict.items()}
+
+
+# ----------------------------------- tests
+
+
+class TestArrayCorrelateFunctions:
```

Comment:
It may just be me not following the pytest syntax, but it's not clear that we are still testing every method (and every type of concurrency) are equal?

Reply:
Ya, it is more concise now but does require a bit of knowledge about pytest fixtures. If you look at the fixture `array_ccs`  it is creating a dictionary. The key is the name of the method and the value is the result of calling the function with the same array_template and array_stream. 

Then `test_single_channel_similar` iterates over all combinations of the values (outputs from the different correlation methods) and makes sure the results are similar. 

I am not sure how you are running the tests but I find myself commenting out the adopts line in pytest.ini, adding `import pdb; pdb.set_trace()` to tests I am interested in looking closer at (for example in the `test_single_channel_similar` function), then running pytest by selecting the node of interest, eg `pytest eqcorrscan/tests/correlate_test.py::TeatArrayCorrelateFunction -s -x --pdb`. This will drop you into a debug prompt to poke around with what is going on in the tests. the s flag will disable output suppression, the x flag tells the tests to stop after one failure, and the pdb flag will drip you into a pdb console whenever a test fails. 

I have also found the pdb++ package very useful. Food for thought.

Reply:
Cool, thanks for the notes - will have to start playing with this style of testing soon - maybe I'll convert some of the other tests slowly to pytest style.

When I'm checking individual tests I usually run:
```python
python <test-file.py> <TestClass>.<test_name>
```
Which is why the `if __name__ == '__main__'` bit was in the tests.

Also, haven't played with pdb, but have seen it mentioned variously, should play.
Thanks for the explanation!